src/execution_plan/ops/op_merge.c

```diff
@@ -257,19 +266,21 @@ static Record MergeConsume(OpBase *opBase) {
 	if(op->bound_variable_stream) OpBase_PropagateFree(op->bound_variable_stream);
 	OpBase_PropagateFree(op->match_stream);
 
-	// If we are setting properties with ON MATCH, execute all pending updates.
-	if(op->on_match && array_len(op->output_records) > 0) _UpdateProperties(op, op->output_records);
 
 	// Exhaust Create stream if we have at least one pattern to create.
 	if(must_create_records) {
 		/* We've populated the Create stream with all the Records it must read;
 		 * pull from it until we've retrieved all newly-created Records. */
+		if(!op->created_records) op->created_records = array_new(Record, 32);
 		Record created_record;
 		while((created_record = _pullFromStream(op->create_stream))) {
-			op->output_records = array_append(op->output_records, created_record);
+			op->created_records = array_append(op->created_records, created_record);
 		}
 	}
 
+	// If we are setting properties with ON MATCH, execute all pending updates.
+	_UpdateProperties(op, op->matched_records);
```

Comment:
Along with the suggestion at 108, `_UpdateProperties` only needs to be called `if (op->on_match)`.

Reply:
```
CREATE (:X)
CREATE (:X)
MERGE (:X)
```
since there is `merge-create` op it is set to be the final writer but it never writes. merge should always be the final writer between the two